### PR TITLE
Implement `sort_remaining` for `sort_index`

### LINKED ISSUE #14011: [FEA] Allow DataFrame.sort_index(level=n) to tiebreak on lexsortedness in mode.pandas_compatable
**Is your feature request related to a problem? Please describe.**
```python
In [1]: import cudf

In [2]:         df = cudf.DataFrame(
   ...:             {
   ...:                 "col1": ["b", "d", "b", "a"],
   ...:                 "col2": [3, 1, 1, 2],
   ...:                 "data": ["one", "two", "three", "four"],
   ...:             }
   ...:         )
   ...: 
   ...:         df2 = df.set_index(["col1", "col2"])

In [3]: df2
Out[3]: 
            data
col1 col2       
b    3       one
d    1       two
b    1     three
a    2      four

In [4]: df2.sort_index(level=0, key=None)
Out[4]: 
            data
col1 col2       
a    2      four
b    3       one
     1     three
d    1       two

In [7]: df2.to_pandas().sort_index(level=0, key=None)
Out[7]: 
            data
col1 col2       
a    2      four
b    1     three
     3       one
d    1       two
```

**Describe the solution you'd like**
The current behavior is still technically correct, but it appears that pandas tiebreaks on lexsortedness. It would be nice to mirror this behavior in pandas compatible mode

**Describe alternatives you've considered**
Manually lexsort after `sort_index`

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

### PR DESCRIPTION
**Pull request by wence-**
## Description

Previously, the `sort_remaining` argument to `sort_index` was ignored. Passing `sort_remaining=False` would raise a `NotImplementedError`. Moreover, for a multiindex, `sort_remaining=True` was not handled correctly: if not all levels were requested as sorted, `sort_index` would behave as if `sort_remaining=False` had been passed.

To fix this case, construct the sort order based on first the provided levels and, if `sort_remaining=True`, the left-over levels (in index order).

To facilitate this, refactor the internal `_get_columns_by_label` function to always return a `Frame`-like object (previously, if we had a `Frame` we would get back a `ColumnAccessor`, and it was only for `IndexedFrame` and above that we'd get something of `Self`-like type back). This meant that calling `_get_sorted_inds` with `by != None` was not possible on an `Index` or `MultiIndex` (the code assumed we'd get a `Frame` back).

- Closes #14011

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/tests/test_multiindex.py [hunk 1]
```diff
@@ -1897,3 +1897,26 @@ def test_multiindex_empty_slice_pandas_compatibility():
...
+            "b": [4, 2, 7, -1, 11, -2, 7, 7],
+            "c": [4, 4, 2, 3, 3, 3, 1, 1],
+            "val": [1, 2, 3, 4, 5, 6, 7, 8],
+        }
+    ).set_index(["a", "b", "c"])
+    cdf = cudf.from_pandas(df)
+
+    expect = df.sort_index(level=levels, sort_remaining=True)
+    got = cdf.sort_index(level=levels, sort_remaining=True)
+    assert_eq(expect, got)
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_multiindex.py:1922]:
```suggestion
)
@pytest.mark.parametrize("sort_remaining", [True, False])
def test_multiindex_sort_index_partial(levels):
    df = pd.DataFrame(
        {
            "a": [3, 3, 3, 1, 1, 1, 2, 2],
            "b": [4, 2, 7, -1, 11, -2, 7, 7],
            "c": [4, 4, 2, 3, 3, 3, 1, 1],
            "val": [1, 2, 3, 4, 5, 6, 7, 8],
        }
    ).set_index(["a", "b", "c"])
    cdf = cudf.from_pandas(df)

    expect = df.sort_index(level=levels, sort_remaining=sort_remaining)
    got = cdf.sort_index(level=levels, sort_remaining=sort_remaining)
    assert_eq(expect, got)
```

Should we just parametrize this test by testing both `sort_remaining=True & False`?

**Review by galipremsagar — APPROVED** [review]
LGTM, minor comment.

File: python/cudf/cudf/tests/test_multiindex.py (hunk 1 above)
Comment (wence-) [reply python/cudf/cudf/tests/test_multiindex.py:1922]: I am not sure if there are any guarantees on matching in that case. (I think whether the sort is stable is an implementation detail).

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_multiindex.py:1922]: Ah, alright. We can ignore this suggestion then. Thanks !

**Review by galipremsagar — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Review by vyasr — APPROVED** [review]

**Pull request merged**